PythonVisualizations/PriorityQueue.py

```diff
@@ -0,0 +1,438 @@
+import random
+from tkinter import *
+
+try:
+    from drawable import *
+    from VisualizationApp import *
+except ModuleNotFoundError:
+    from .drawable import *
+    from .VisualizationApp import *
+
+
+class PriorityQueue(VisualizationApp):
+    CELL_SIZE = 50
+    CELL_BORDER = 2
+    CELL_BORDER_COLOR = 'black'
+    ARRAY_X0 = 100
+    ARRAY_Y0 = 100
+
+    colors = ['red', 'green', 'blue', 'orange', 'yellow', 'cyan', 'magenta',
+              'dodgerblue', 'turquoise', 'grey', 'gold', 'pink']
+    nextColor = 0
+
+    def __init__(self, size=12, title="Priority Queue", **kwargs):
+        super().__init__(title=title, **kwargs)
+        self.size = size
+        self.title = title
+        self.list = []
+        self.buttons = self.makeButtons()
+
+        
+        self.display()
+
+    def __str__(self):
+        return str(self.list)
+
+    # ARRAY FUNCTIONALITY
+
+    def createIndex(self, index,  # cell
+                    name=None,  # with an optional name label
+                    level=1,  # at a particular level away from the cells
+                    color=None):  # (negative are below)
+        if not color: color = self.VARIABLE_COLOR
+
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        level_spacing = self.VARIABLE_FONT[1]
+        x = cell_center[0]
+        if level > 0:
+            y0 = cell_coords[1] - self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[1] - self.CELL_SIZE * 3 // 10
+        else:
+            y0 = cell_coords[3] + self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[3] + self.CELL_SIZE * 3 // 10
+        arrow = self.canvas.create_line(
+            x, y0, x, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x + 2, y0, text=name, anchor=SW if level > 0 else NW,
+                font=self.VARIABLE_FONT, fill=color)
+        return (arrow, label) if name else (arrow,)
+
+    def insert(self, val):
+        callEnviron = self.createCallEnvironment()
+        self.startAnimations()
+        j = self.find(val)  # Find where item should go
+
+        self.list.append(drawable(None))
+        
+        indexK = self.createIndex(len(self.list) -1, 'k', level=-2) # create "k" arrow
+        callEnviron |= set(indexK)  
+        
+        for k in range(len(self.list) - 1, j, -1):  # Move bigger items right
+            
+            self.wait(1)
+            for item in indexK:
+                self.canvas.move(item, -self.CELL_SIZE, 0)  # Move "k" arrow  
+                
+            self.list[k].val = self.list[k - 1].val 
+            self.assignElement(k - 1, k, callEnviron)
```

Comment:
The shift loop code is
```
      j = self.__nItems - 1           # Start at front
      while j >= 0 and (              # Look for place by priority
            self.__pri(item) >= self.__pri(self.__que[j])):
         self.__que[j+1] = self.__que[j] # Shift items to front
         j -= 1                       # Step towards rear
```
Please model this loop on that.  The `indexK` should be `indexJ`.  You can assume the `self.__pri()` function is the identity function for this visualization.

Reply:
I'm stuck on this one - the code keeps returning (mostly tkinter-related) errors when I try to insert anything.

Reply:
I tried checking out what you have committed to the branch (through commit 102dd99).  When I try inserting items in that version, it works and doesn't throw any exceptions.

Reply:
When I try to insert items, it gives me this:
```
  File "c:\Users\Natania\AppData\Local\Programs\Python\Python38-32\lib\tkinter\__init__.py", line 1883, in __call__
    return self.func(*args)
  File "C:\Users\Natania\Documents\GitHub\datastructures-visualization\PythonVisualizations\VisualizationApp.py", line 256, in animatedOperation
    self.cleanUp()
  File "x-wingide-python-shell://119930528/2", line 296, in cleanUp
  File "x-wingide-python-shell://119930528/2", line 290, in fixCells
AttributeError: 'NoneType' object has no attribute 'display_shape'

```
It seems that there is some reference to an empty array slot where there should be a reference to a drawable. The error comes from VizApp instead of PQ.

Reply:
I agree.  The `fixCells()` expects `self.list` to have only drawables in it.  Here's a suggestion if your text editor supports pattern search.  Look in the code for every line like `self.list[.*] = ` where ".*" means any sequence of characters.  Somewhere, something on the right side of the assignment is getting set to `None`.

Reply:
When I look in my version or PriorityQueue.py at line 91 I see this:
```
        self.list[j]= (drawable(
            val, self.canvas.itemconfigure(cellPair[0], 'fill'), *cellPair))
```
The outer set of parentheses isn't needed, but shouldn't prevent this from storing a `drawable` in cell `j`.  There is a problem in that `itemconfigure()` returns a tuple and its the last element of the tuple that has the color.  It should be `self.canvas.itemconfigure(cellPair[0], 'fill')[-1]` to extract that last element.

There are also a couple of `self.list.append()` calls.  At line 67 at the beginning of `insert()` I see:
```
        self.list.append(drawable(None))
```
That makes an empty drawable with all fields filled with `None`, but there should still be `display_shape` field.  I'm not sure if it's a good idea to append the empty drawable at this point.  I think that empty drawable is needed for the call to `assignElement()` at line 79.  What's odd in my version is that it has:
```
            self.list[k].val = self.list[k - 1].val 
            self.assignElement(k - 1, k, callEnviron)
```
The assignment of `.val` is done before calling `assignElement()`, which re-assigns `.val` as part of its body.  That shouldn't cause the error you mentioned, but it seems redundant.

Reply:
I switched the order of the ```assignElement()``` call and the ```.val``` assignment, but nothing changed. Also, ```self.list.append(drawable(None))``` was changed to ```self.list[self.nItems] = drawable(None)``` a few commits back.